Replace external annotation service with internal JSON→Cypher pipeline and add general query support

dagemawinegash opened the pull request:
### Overview
This PR removes the dependency on the external `annotation_service` URL and replaces it with a fully internal pipeline that:
- Extracts structured JSON from user queries using updated prompts
- Converts validated JSON into Cypher with an internal converter
- Executes Cypher directly against Neo4j
- Summarizes results into user-friendly text
- Unifies routing through `assistant()` so history is consistently stored
- Adds support for general database queries (statistics, counts, overviews) alongside biological entity queries
---
## Files changed

### Updated

`AI-Assistant/app/prompts/annotation_prompts.py` – Adds ID vs property rules, relationship guidance, and summarization prompts.  

`AI-Assistant/app/prompts/classifier_prompt.py` – Added main classifier prompt for broader query classification (biological vs general queries).

`AI-Assistant/app/annotation_graph/annotated_graph.py` – Adds Cypher execution, result summarization, full annotation pipeline, and general query handling with database summary approach.  

`AI-Assistant/app/main.py` – Routes annotation queries through unified pipeline, supports both biological and general queries, and returns detailed outputs.  

`AI-Assistant/app/__init__.py` – Updated to use new enhanced schema file for improved schema management.

`AI-Assistant/config/config.yaml` – Updated schema path configuration to point to new schema files.

`AI-Assistant/docker-compose.yml` – Configures Neo4j 5.15 with APOC 

### Created
`AI-Assistant/app/annotation_graph/json_to_cypher.py` – Converts validated JSON into safe Cypher queries with explicit relationships.

`AI-Assistant/config/new_enhanced_schema.txt` – Simplified schema with `gene`, `transcript`, `exon` nodes and main relationships, aligned with CSVs.  

`AI-Assistant/config/new_preprocessed_schema.txt` – New preprocessed schema file aligned with the enhanced schema structure.

---

## Checklist
- [x] Remove dependency on external `annotation_service` URL
- [x] Internal JSON→Cypher converter with sanitization
- [x] Execute and summarize results via internal pipeline
- [x] Unify routing through `assistant()` for consistent history
- [x] Update prompts and schemas to the minimal GENCODE model
- [x] Add support for general database queries (statistics, counts, overviews)
- [x] Implement query classification to route biological vs general queries appropriately
- [x] Preserve legacy schema files for reference and comparison

kewserseid committed:
Merge pull request #37 from rejuve-bio/testing

Testing

kewserseid committed:
Merge pull request #38 from rejuve-bio/testing

quick fixes

dagemawinegash committed:
Simplify BioCypher schemas to match minimal GENCODE model
- Keep only gene, transcript, exon nodes and key relationships
- Clarify ID vs properties with concrete examples
- Align property names with imported CSVs

dagemawinegash committed:
Created Neo4j with APOC plugin
- Use neo4j:5.15 (community) with APOC enabled
- Allow file import via apoc settings

dagemawinegash committed:
Revise annotation prompts with ID/property rules and summarization
- Add CRITICAL ID vs PROPERTIES rules with examples
- Add relationship inference rules for predicates
- Introduce RESULT_SUMMARIZATION_PROMPT

dagemawinegash committed:
Add JsonToCypherConverter to replace external API
- Build MATCH/WHERE/RETURN with relationship variables
- Validate JSON shape and sanitize values
- Support predicates and selective returns

dagemawinegash committed:
Internalize annotation pipeline: execute Cypher and summarize
- Add execute_cypher_query and summarize_results
- Implement process_annotation_query for method orchestration
- Remove legacy API methods and unused imports
- Return structured pipeline_status and results

dagemawinegash committed:
Integrate internal annotation agent and route via assistant()
- Call process_annotation_query in annotation_agent
- Preserve pipeline_details in final response
- Remove unused get_json_format tool wiring

kewserseid commented on config/enhanced_schema.txt:
Lets comment them rather than removing... since you will use them when you will be adding the rest datas.

dagemawinegash replied:
yes we can do that but since the current files are txt files we can't comment out lines like code files. there are other methods that we can use like "if line.startswith("#"): continue" by iterating over each line but instead of doing that I think it is better to use the current for now and when we need the previous one we can refer it since we have the code on github. what do you think?

kewserseid replied:
you can create another file that is fine.

dagemawinegash replied:
done!

kewserseid commented on config/preprossed_schema.txt:
same here.

dagemawinegash replied:
we can use the same appraoch here too.

kewserseid commented on docker-compose.yml:
can't we call the database through url? what do you think? I dont think it need to be on docker compose as an image. since we will be using the deployed one lets assume we can access it through url. and just store the values on env file and calling it would be enough form my view. for testing purpose we can just use our local neo4j and call that through url.

dagemawinegash replied:
yes you are right. I have updated it.

dagemawinegash committed:
Merge branch 'rejuve-bio:main' into update_annotation

dagemawinegash committed:
Removed the neo4j instance
- because we need it for testing purposes only so no need to run it on docker

dagemawinegash committed:
added the json query on the response for visualization need from the frontend

dagemawinegash committed:
Add main classifier prompt for broader query classification

- Add main_classifier_prompt to classify queries into annotation_biological, annotation_general, hypothesis, galaxy, and rag categories

dagemawinegash committed:
Integrate general annotation queries with existing biological pipeline

- Update _classify_query to use main_classifier_prompt for broader query classification
- Add annotation_general routing to annotation_agent in workflow creation
- Modify _annotation_agent to pass query_type parameter to process_annotation_query method

dagemawinegash committed:
Implement general query handling with database summary approach

- Add process_annotation_query method to route between biological and general query handling
- Implement _handle_general_query and _generate_database_summary for database statistics queries

kewserseid commented on app/annotation_graph/annotated_graph.py:
can you replace the annotation code here?

dagemawinegash replied:
I already replaced the annotation in the "annotation_graph.py" with the ideas that we discussed about. 
do you want me to change anything else?

kewserseid replied:
i meant the previous. lets not delete them. rather add additional methods you have added. store the previous one as well but use the new imeplementation.

dagemawinegash replied:
done!

kewserseid commented on docker-compose.yml:
is the json query format or the general summary takes too long to execute?

dagemawinegash replied:
the quick fix that we discussed on(returning both the json and the summary for both the general and biological queries) was generating cypher queries that take too much time to get executed for general queries only(not for biological queries). But now this issue is fixed as I have implemented the dual-pipeline that we discussed on so now everyting is fixed so nothing takes too long now.

dagemawinegash committed:
Restore legacy code in annotated_graph for comparison
- Re-add previously removed sections in app/annotation_graph/annotated_graph.py for future reference

dagemawinegash committed:
Add new schema definitions
- Mirrors current minimal schema and rules; kept separate for safe comparison

dagemawinegash committed:
Point app to new enhanced schema file
- Update app initialization to read new_enhanced_schema.txt
- Update config.yaml SCHEMA_PATH to the new enhanced schema

dagemawinegash committed:
Preserve legacy schema files as reference
- Keep enhanced_schema.txt and preprossed_schema.txt unchanged for backward reference
- Clarifies they’re legacy and not the current runtime source

dagemawinegash committed:
Fix classifier prompt to distinguish count queries from specific entity queries
- Update annotation_biological to focus on "find/retrieve specific entities" rather than just mentioning biological terms
- Update annotation_general to emphasize "aggregate statistics and database metadata" to properly classify count queries

kewserseid commented:
small improvement on the SentenceTransformer model is to not load inside the function every time it’s called. This can be inefficient if the function is called multiple times.
we can load the model once at the top-level, outside the function, and then just use it inside the function. For example:

model = SentenceTransformer("all-MiniLM-L6-v2")
def sentence_transformer_embedding_model(batch):
       conitnue the rest of the code here ...

dagemawinegash committed:
Load the SentenceTransformer model once at module level
- Refactor LLMModel to use the preloaded model for efficiency in multiple calls

dagemawinegash commented:
yes, you are right. 
I have updated the code so that the SentenceTransformer model is loaded once.
